contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,312 @@
+         */
+    }
+
+    IERC20Upgradeable public joe;
+
+    /// @notice Internal balance of $JOE, this gets updated on user deposits / withdrawals
+    /// @dev this allows to reward users with $JOE.
+    uint256 internalJoeBalance;
+    /// @notice The array of token that people can claim to get reward
+    IERC20Upgradeable[] public rewardTokens;
+    mapping(IERC20Upgradeable => bool) public isRewardToken;
+    /// @notice Last reward balance of `token`
+    mapping(IERC20Upgradeable => uint256) public lastRewardBalance;
+
+    address public feeCollector;
+
+    /// @notice The deposit fee, scaled to `PRECISION`
+    uint256 public depositFeePercent;
+
+    /// @notice Accumulated `token` rewards per share, scaled to `PRECISION`. See above
+    mapping(IERC20Upgradeable => uint256) public accRewardPerShare;
+    /// @notice `PRECISION` of `accRewardPerShare`
+    uint256 public PRECISION;
+
+    /// @notice Info of each user that stakes JOE
+    mapping(address => UserInfo) private userInfo;
+
+    /// @notice Emitted when a user deposits JOE
+    event Deposit(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user withdraws JOE
+    event Withdraw(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user emergency withdraws its JOE
+    event EmergencyWithdraw(address indexed user, uint256 amount);
+
+    /**
+     * @notice Initialize a new StableJoeStaking contract
+     * @dev This contract needs to receive an ERC20 `_rewardToken` in order to distribute them
+     * (with JoeMakerV4 in our case)
+     * @param _rewardToken The address of the ERC20 reward token
+     * @param _joe The address of the JOE token
+     * @param _depositFeePercent The deposit fee percent, scalled to 1e18, e.g. 3% is 3e16
+     */
+    function initialize(
+        IERC20Upgradeable _rewardToken,
+        IERC20Upgradeable _joe,
+        address _feeCollector,
+        uint256 _depositFeePercent
+    ) external initializer {
+        __Ownable_init();
+        require(_feeCollector != address(0), "StableJoeStaking: fee collector can't be address 0");
+        require(_depositFeePercent < 5e17, "StableJoeStaking: max deposit fee can't be greater than 50%");
+
+        joe = _joe;
+        depositFeePercent = _depositFeePercent;
+        feeCollector = _feeCollector;
+
+        isRewardToken[_rewardToken] = true;
+        rewardTokens.push(_rewardToken);
+        PRECISION = 1e24;
+    }
+
+    /**
+     * @notice Deposit JOE for reward token allocation
+     * @param _amount The amount of JOE to deposit
+     */
+    function deposit(uint256 _amount) external {
+        UserInfo storage user = userInfo[msg.sender];
+
+        uint256 fee = (_amount * depositFeePercent) / 1e18;
+        uint256 amountMinusFee = _amount.sub(fee);
+
+        uint256 previousAmount = user.amount;
+        uint256 newAmount = user.amount.add(amountMinusFee);
+        user.amount = newAmount;
+
+        uint256 len = rewardTokens.length;
```

Comment:
Any particular reason for this variable `len` instead of just using `i < rewardTokens.length` in the for loop. I don't see it used anywhere else.

Reply:
That's intended, when reading a value from the memory it costs more gas than when you use a var that is created inside the function, and as we use it in in a loop it's even better to take care of gas here :)

Reply:
TIL